pkg/bootkube/create.go

```diff
@@ -205,24 +211,175 @@ func customResourceDefinitionKindURI(apiGroup, version, namespace, plural string
 		strings.ToLower(plural))
 }
 
-func apiTest(c clientcmd.ClientConfig) error {
-	config, err := c.ClientConfig()
+func (c *creater) create(m manifest) error {
+	plural, err := c.mapper.resourcePlural(m.apiVersion, m.kind)
 	if err != nil {
-		return err
+		return fmt.Errorf("dicovery failed: %v", err)
 	}
-	client, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return err
+
+	err = c.client.Post().
+		AbsPath(m.urlPath(plural)).
+		Body(m.raw).
+		SetHeader("Content-Type", "application/json").
+		Do().Error()
+	if err != nil && errors.IsAlreadyExists(err) {
+		// retry with a PUT
+		err = c.client.Put().
+			AbsPath(m.urlPath(plural)+"/"+m.name).
+			Body(m.raw).
+			SetHeader("Content-Type", "application/json").
+			Do().Error()
 	}
+	return err
+}
 
-	// API Server is responding
-	healthStatus := 0
-	client.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
-	if healthStatus != http.StatusOK {
-		return fmt.Errorf("API Server http status: %d", healthStatus)
+func (m manifest) urlPath(plural string) string {
+	u := "/apis"
+	if m.apiVersion == "v1" {
+		u = "/api"
 	}
+	u = u + "/" + m.apiVersion
+	if m.namespace != "" {
+		u = u + "/namespaces/" + m.namespace
+	}
+	return u + "/" + plural
+}
 
-	// System namespace has been created
-	_, err = client.Namespaces().Get(api.NamespaceSystem, metav1.GetOptions{})
-	return err
+// loadManifests parses a directory of YAML Kubernetes manifest.
+func loadManifests(p string) ([]manifest, error) {
+	var manifests []manifest
+	err := filepath.Walk(p, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			b := filepath.Base(p)
+			if b != p && strings.HasPrefix(b, ".") {
+				// Ignore directories that start with a "."
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("open %s: %v", path, err)
+		}
+		defer f.Close()
+
+		ms, err := parseManifests(f)
+		if err != nil {
+			return fmt.Errorf("parse file %s: %v", path, err)
+		}
+		for i := range ms {
+			ms[i].filepath = path
+		}
+		manifests = append(manifests, ms...)
+		return nil
+	})
+	return manifests, err
+}
+
+// parseManifests parses a YAML or JSON document that may contain one or more
+// kubernetes resoures.
+func parseManifests(r io.Reader) ([]manifest, error) {
+	reader := yaml.NewYAMLReader(bufio.NewReader(r))
+	var manifests []manifest
+	for {
+		data, err := reader.Read()
+		if err != nil {
+			if err == io.EOF {
```

Comment:
this seems odd to me... usually we would want to read all the chunks until EOF, then parse the data.

Reply:
`NewYAMLReader` is a little special. It's not a stream implementation, it returns discrete chunks of YAML https://godoc.org/k8s.io/apimachinery/pkg/util/yaml#YAMLReader.Read

Maybe renaming `data` to `rawManifest` would make that clearer?

The following Go idiom is also pretty common:

```go
if err != nil {
    if err == io.EOF {
        // successful return
    }
    // unsuccessful return
}
```

Take a look at the Go std library:

```
grep -r '== io.EOF' -A8 "$(go env GOROOT)/src/" | grep -v '_test.go' | less
```

Reply:
gotcha. Thanks for the explanation. parseManifests seems fine.